### DISCUSSION TITLE
LEMMA pattern doesn't work in Attribute Ruler - spaCy 3.0.5

### DISCUSSION THREAD

**Discussion by MausamGaurav** [explosion/spaCy#7904] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

I perform the below. The attribute ruler should pick the pattern that whenever the token of a word has 'LEMMA': 'review', it should always be shown as verb and not noun.
```
nlp = spacy.load('en_core_web_sm')
ruler = nlp.get_pipe("attribute_ruler")
patterns = [[{"LEMMA": "review"}]]
attrs = {"POS": "VERB"}
ruler.add(patterns=patterns, attrs=attrs)
```
```
test = 'review email'
doc = nlp(test)
for token in doc:
    print(token.lemma_, token.pos_)
```
However the pattern doesn't appear to work and the output is noun.
```
review NOUN
email NOUN
```

If I change the pattern to 'LOWER', it works.
Code:
```
nlp = spacy.load('en_core_web_sm')
ruler = nlp.get_pipe("attribute_ruler")
patterns = [[{"LOWER": "review"}]]
attrs = {"POS": "VERB"}
ruler.add(patterns=patterns, attrs=attrs)
```

```
test = 'review email'
doc = nlp(test)
for token in doc:
    print(token.lemma_, token.pos_)
```
Output:

```
review VERB
email NOUN
```

## Your Environment

* Operating System: Windows 10
* Python Version Used: 3.7.4

**Comment by adrianeboyd** [comment, marked as answer]
Hi, this behavior is related to the order of components in the pipeline. The lemmatizer hasn't run when the attribute ruler runs, so the lemmas aren't set yet. In many pipelines (it depends on the language, but English is one of them), the attribute ruler may be required for the lemmatizer to work because the attribute ruler sets `POS` for the lemmatizer (based on the tags from the tagger). See here for more details: https://spacy.io/models#design

To apply the rule above, you can add a second attribute ruler with a custom name anywhere after the lemmatizer in the pipeline (here this adds it last):

```python
nlp.add_pipe("attribute_ruler", name="attribute_ruler_on_lemmas")
```
